### ISSUE TITLE
[ENHANCEMENT] Adaptive Softmax

### ISSUE THREAD

**Issue by siddsach** [allenai/allennlp#981]
Hey!

Thanks for this awesome Pytorch project! I'm totally sold on the idea, and wanted to take a crack and getting involved! 

I found recently that for language model training where the output vocabulary space is high-dimensional, that Adaptive Softmax from (https://arxiv.org/abs/1609.04309) can lead to a huge speed up. It's essentially a modification of hierarchical softmax that is optimized for GPU computing, and it can sometimes lead to orders of magnitude speed improvements. I figured that since this would make seq2seq decoding a lot faster, it'd perhaps be useful, not to mention a good way for me to dive in and find my way around. Does that sound helpful? If not a have a few other features I want to implement so I could propose something else. 

If that sounds good, I got the code from this repo: https://github.com/rosinality/adaptive-softmax-pytorch, to work with the word_language_model example in Pytorch, and thought it'd be a good place to start. I figured that it'd make sense to implement the output projection, activation, and loss in an nn module, and then add an option for seq2seq training to use this activation and loss rather than the normal seq2seq and crossentropy. Does that sound good?

Also I noticed in the seq2seq training file that the decoder is hardcoded as an LSTM, but is intended to implemented differently by @pdasigi. If you don't mind my asking, what's the plan for that? I want to use the Transformer seq2seq which requires changing the decoder.

**Comment by matt-gardner** [comment]
Yeah, that sounds like a great thing to work on.  We haven't put a whole lot of effort into optimizing our seq2seq model, or using it to actually train a useful model on real data.  If you want to improve our seq2seq code, great!  We'd love to include a nice seq2seq model on some task in our demo, we just don't have the bandwidth to pursue that ourselves.

**Comment by pdasigi** [comment]
Regarding the comment in code, it refers to making the decoder configurable and reading its configuration from the params file as well, instead of hardcoding it as an LSTM. As Matt said, we do not have immediate plans for doing it, and we'd love it if you want to contribute.

**Comment by siddsach** [comment]
Awesome! I'll work on both adasoft and the decoder

**Comment by nicolabertoldi** [comment]
I am very interested in having adaptive_softmax in the transformer decoder.

Which is the status of your work?
